source/common/config/incremental_subscription_impl.h

```diff
@@ -0,0 +1,223 @@
+#pragma once
+
+#include <queue>
+
+#include "envoy/api/v2/discovery.pb.h"
+#include "envoy/common/token_bucket.h"
+#include "envoy/config/subscription.h"
+
+#include "common/common/assert.h"
+#include "common/common/backoff_strategy.h"
+#include "common/common/logger.h"
+#include "common/common/token_bucket_impl.h"
+#include "common/config/grpc_stream.h"
+#include "common/config/utility.h"
+#include "common/grpc/common.h"
+#include "common/protobuf/protobuf.h"
+#include "common/protobuf/utility.h"
+
+namespace Envoy {
+namespace Config {
+
+struct ResourceNameDiff {
+  std::vector<std::string> added_;
+  std::vector<std::string> removed_;
+};
+
+/**
+ * Manages the logic of a (non-aggregated) incremental xDS subscription.
+ */
+template <class ResourceType>
+class IncrementalSubscriptionImpl
+    : public Subscription<ResourceType>,
+      public GrpcStream<envoy::api::v2::IncrementalDiscoveryRequest,
+                        envoy::api::v2::IncrementalDiscoveryResponse, ResourceNameDiff> {
+public:
+  IncrementalSubscriptionImpl(const LocalInfo::LocalInfo& local_info,
+                              Grpc::AsyncClientPtr async_client, Event::Dispatcher& dispatcher,
+                              const Protobuf::MethodDescriptor& service_method,
+                              Runtime::RandomGenerator& random, Stats::Scope& scope,
+                              const RateLimitSettings& rate_limit_settings, SubscriptionStats stats)
+      : GrpcStream<envoy::api::v2::IncrementalDiscoveryRequest,
+                   envoy::api::v2::IncrementalDiscoveryResponse, ResourceNameDiff>(
+            std::move(async_client), service_method, random, dispatcher, scope,
+            rate_limit_settings),
+        type_url_(Grpc::Common::typeUrl(ResourceType().GetDescriptor()->full_name())),
+        local_info_(local_info), stats_(stats) {
+    request_.set_type_url(type_url_);
+    request_.mutable_node()->MergeFrom(local_info_.node());
+  }
+
+  // Enqueues and attempts to send a discovery request, (un)subscribing to resources missing from /
+  // added to the passed 'resources' argument, relative to resources_. Updates resources_ to
+  // 'resources'.
+  void buildAndQueueDiscoveryRequest(const std::vector<std::string>& resources) {
+    ResourceNameDiff diff;
+    std::set_difference(resources.begin(), resources.end(), resource_names_.begin(),
+                        resource_names_.end(), diff.added_.begin());
```

Comment:
I get a segfault here, had to use an insert iterator to get this to work:
```
std::set_difference(resources.begin(), resources.end(), resource_names_.begin(),
    resource_names_.end(), std::inserter(diff.added_, diff.added_.begin()));
```

Reply:
Done.